core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -292,56 +286,44 @@ class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: String,
    */
   private class Watchers(val key: Any) {
 
-    private[this] val operations = new LinkedList[T]()
+    private[this] var operations = Vector[T]()
 
-    def watched: Int = operations synchronized operations.size
+    def watched: Int = synchronized(operations.size)
 
     // add the element to watch
     def watch(t: T) {
-      operations synchronized operations.add(t)
+      synchronized {
+        operations = operations :+ t
+      }
     }
 
     // traverse the list and try to complete some watched elements
     def tryCompleteWatched(): Int = {
-
-      var completed = 0
-      operations synchronized {
-        val iter = operations.iterator()
-        while (iter.hasNext) {
-          val curr = iter.next()
-          if (curr.isCompleted) {
-            // another thread has completed this operation, just remove it
-            iter.remove()
-          } else if (curr synchronized curr.tryComplete()) {
-            completed += 1
-            iter.remove()
-          }
-        }
+      // call tryComplete without holding the lock to avoid potential deadlocks
+      var completedAlready = 0
+      var completedNow = 0
+      for (op <- synchronized(operations)) {
```

Comment:
Adding to @becketqin's comment on readability. It might be more immediately obvious what this is doing by changing the loop to:
```scala
val ops = synchronized {
  operations
}
for (op <- ops)
```